### ISSUE TITLE
Does not detect drive serial numbers.

### ISSUE THREAD

**Issue by azilber** [rockstor/rockstor-core#2316]
I have 4, 6TB HGST drives attached via USB Dock.  The drives have identical model numbers, but different serial numbers.   Rockstor doesn't detect the serial numbers:
![image](https://user-images.githubusercontent.com/691270/134774036-cea15e67-69ae-4ee3-82eb-af57b199d1f1.png)

Yet:

> winnuc:~ # smartctl -i /dev/sdb |grep Serial
> Serial Number:    K1GZ601D
> winnuc:~ # smartctl -i /dev/sdc |grep Serial
> Serial Number:    K1KZX3PD
> winnuc:~ # smartctl -i /dev/sdd |grep Serial
> Serial Number:    K1GYXAXD
> winnuc:~ # smartctl -i /dev/sde |grep Serial
> Serial Number:    NCH3GGBV

Clicking on the drive ID also shows:

>             Traceback (most recent call last):
>   File "/opt/rockstor/src/rockstor/storageadmin/views/disk_smart.py", line 58, in _validate_disk
>     return Disk.objects.get(id=did)
>   File "/opt/rockstor/eggs/Django-1.8.16-py2.7.egg/django/db/models/manager.py", line 127, in manager_method
>     return getattr(self.get_queryset(), name)(*args, **kwargs)
>   File "/opt/rockstor/eggs/Django-1.8.16-py2.7.egg/django/db/models/query.py", line 334, in get
>     self.model._meta.object_name
> DoesNotExist: Disk matching query does not exist.
> 
>

**Comment by phillxnet** [comment]
@azilber @Thanks for the report. This is a know hardware issue with some USB external multi-device enclosures. It is a 'feature' of the hardware and renders that hardware incompatible with udev's reading of the serial numbers and thus Rockstor's use of those enclosures. Take a look at the following developer wiki entry on our forum:

"Device management in Rockstor": https://forum.rockstor.com/t/device-management-in-rockstor/1768



> Yet:
> 
> > winnuc:~ # smartctl -i /dev/sdb |grep Serial
> > Serial Number:    K1GZ601D
> > winnuc:~ # smartctl -i /dev/sdc |grep Serial
> > Serial Number:    K1KZX3PD
> > winnuc:~ # smartctl -i /dev/sdd |grep Serial
> > Serial Number:    K1GYXAXD
> > winnuc:~ # smartctl -i /dev/sde |grep Serial
> > Serial Number:    NCH3GGBV

Smartmontools retrieves the drives serial numbers via it's own drive interigation. We currently rely only on udev which in turn creates the required by-id device names that we also rely upon.

To confirm this what are the outputs of the following command when passed each drives device name:

```
udevadm info --name=device_name
```
Just retrieving the device serial is not the whole story. Udev does this and then, in turn, creates the by-id names. So even if we failed over to smartmontools (as our second fail over incidentally) we would still be without the by-id names as udev fails to make them. Again confirmation of this would be good in our case by pasting in the output from the following command, just in case this is not the suspected hardware 'feature' of assigning the same serial to all members within it's enclosure/remit.

```
ls -la /dev/disk/by-id/
```

It would also be most useful to have the exact make and model and firmware version, if available, of this external enclosure so we can keep a tally of such enclosures and label them as incompatible. The simply mis-represent / obfuscate the only direct hardware identifier we have to identify individual disks reliably.

In another instance/report I believe we had an enclosure that would report a 'made-up' serial for each slot. Again this defeats udev's creation of by-id names (as it uses serial number) and consequently our ability to 'track' drives by their serial numbers.

**Comment by azilber** [comment]
The enclosure is "Mediasonic 4 Bay Dock for 2.5' / 3.5' SATA HDD / SSD - USB 3.0 & eSATA (HFD1-SU3S2)".  It's connected over USB 3.0 to an Intel NUC.   The drives I pulled from my old Helios4 NAS, which have an md RAID-5 with btrfs over it.  I imported the pool into Rockstor.

lsusb reports it as:

> Bus 002 Device 002: ID 152d:0567 JMicron Technology Corp. / JMicron USA Technology Corp. JMS567 SATA 6Gb/s bridge
> Device Descriptor:
>   bLength                18
>   bDescriptorType         1
>   bcdUSB               3.00
>   bDeviceClass            0
>   bDeviceSubClass         0
>   bDeviceProtocol         0
>   bMaxPacketSize0         9
>   idVendor           0x152d JMicron Technology Corp. / JMicron USA Technology Corp.
>   idProduct          0x0567 JMS567 SATA 6Gb/s bridge
>   bcdDevice            2.05
>   iManufacturer          10 JMicron
>   iProduct               11 USB to ATA/ATAPI Bridge
>   iSerial                 5 152D00539000
>   bNumConfigurations      1

Here's the requested udev info:

winnuc:/var/log #  udevadm info -n /dev/sdb

> P: /devices/pci0000:00/0000:00:14.0/usb2/2-1/2-1:1.0/host2/target2:0:0/2:0:0:0/block/sdb
> N: sdb
> S: disk/by-id/usb-HGST_HDN_726060ALE614_152D00539000-0:0
> S: disk/by-path/pci-0000:00:14.0-usb-0:1:1.0-scsi-0:0:0:0
> E: COMPAT_SYMLINK_GENERATION=2
> E: DEVLINKS=/dev/disk/by-path/pci-0000:00:14.0-usb-0:1:1.0-scsi-0:0:0:0 /dev/disk/by-id/usb-HGST_HDN_726060ALE614_152D00539000-0:0
> E: DEVNAME=/dev/sdb
> E: DEVPATH=/devices/pci0000:00/0000:00:14.0/usb2/2-1/2-1:1.0/host2/target2:0:0/2:0:0:0/block/sdb
> E: DEVTYPE=disk
> E: DONT_DEL_PART_NODES=1
> E: ID_BUS=usb
> E: ID_FS_LABEL=helios4:Main
> E: ID_FS_LABEL_ENC=helios4:Main
> E: ID_FS_TYPE=linux_raid_member
> E: ID_FS_USAGE=raid
> E: ID_FS_UUID=0c53ed8f-2ee6-c022-a0be-c9c7307a30bb
> E: ID_FS_UUID_ENC=0c53ed8f-2ee6-c022-a0be-c9c7307a30bb
> E: ID_FS_UUID_SUB=0d6c37ce-11c3-945d-7878-527140aeb0f3
> E: ID_FS_UUID_SUB_ENC=0d6c37ce-11c3-945d-7878-527140aeb0f3
> E: ID_FS_VERSION=1.2
> E: ID_INSTANCE=0:0
> E: ID_MODEL=726060ALE614
> E: ID_MODEL_ENC=726060ALE614\x20\x20\x20\x20
> E: ID_MODEL_ID=0567
> E: ID_PATH=pci-0000:00:14.0-usb-0:1:1.0-scsi-0:0:0:0
> E: ID_PATH_TAG=pci-0000_00_14_0-usb-0_1_1_0-scsi-0_0_0_0
> E: ID_REVISION=0125
> E: ID_SERIAL=HGST_HDN_726060ALE614_152D00539000-0:0
> E: ID_SERIAL_SHORT=152D00539000
> E: ID_TYPE=disk
> E: ID_USB_DRIVER=usb-storage
> E: ID_USB_INTERFACES=:080650:
> E: ID_USB_INTERFACE_NUM=00
> E: ID_VENDOR=HGST_HDN
> E: ID_VENDOR_ENC=HGST\x20HDN
> E: ID_VENDOR_ID=152d
> E: MAJOR=8
> E: MD_DEVICE=md127
> E: MD_DEVNAME=helios4:Main
> E: MD_FOREIGN=yes
> E: MD_STARTED=no
> E: MINOR=16
> E: SUBSYSTEM=block
> E: TAGS=:systemd:
> E: USEC_INITIALIZED=553844851

winnuc:/var/log # udevadm info -n /dev/sdc

> P: /devices/pci0000:00/0000:00:14.0/usb2/2-1/2-1:1.0/host2/target2:0:0/2:0:0:1/block/sdc
> N: sdc
> S: disk/by-id/usb-HGST_HDN_726060ALE614_152D00539000-0:1
> S: disk/by-path/pci-0000:00:14.0-usb-0:1:1.0-scsi-0:0:0:1
> E: COMPAT_SYMLINK_GENERATION=2
> E: DEVLINKS=/dev/disk/by-id/usb-HGST_HDN_726060ALE614_152D00539000-0:1 /dev/disk/by-path/pci-0000:00:14.0-usb-0:1:1.0-scsi-0:0:0:1
> E: DEVNAME=/dev/sdc
> E: DEVPATH=/devices/pci0000:00/0000:00:14.0/usb2/2-1/2-1:1.0/host2/target2:0:0/2:0:0:1/block/sdc
> E: DEVTYPE=disk
> E: DONT_DEL_PART_NODES=1
> E: ID_BUS=usb
> E: ID_FS_LABEL=helios4:Main
> E: ID_FS_LABEL_ENC=helios4:Main
> E: ID_FS_TYPE=linux_raid_member
> E: ID_FS_USAGE=raid
> E: ID_FS_UUID=0c53ed8f-2ee6-c022-a0be-c9c7307a30bb
> E: ID_FS_UUID_ENC=0c53ed8f-2ee6-c022-a0be-c9c7307a30bb
> E: ID_FS_UUID_SUB=78170a3e-15ec-05e8-a7b9-0edf4ddecf0a
> E: ID_FS_UUID_SUB_ENC=78170a3e-15ec-05e8-a7b9-0edf4ddecf0a
> E: ID_FS_VERSION=1.2
> E: ID_INSTANCE=0:1
> E: ID_MODEL=726060ALE614
> E: ID_MODEL_ENC=726060ALE614\x20\x20\x20\x20
> E: ID_MODEL_ID=0567
> E: ID_PATH=pci-0000:00:14.0-usb-0:1:1.0-scsi-0:0:0:1
> E: ID_PATH_TAG=pci-0000_00_14_0-usb-0_1_1_0-scsi-0_0_0_1
> E: ID_REVISION=0125
> E: ID_SERIAL=HGST_HDN_726060ALE614_152D00539000-0:1
> E: ID_SERIAL_SHORT=152D00539000
> E: ID_TYPE=disk
> E: ID_USB_DRIVER=usb-storage
> E: ID_USB_INTERFACES=:080650:
> E: ID_USB_INTERFACE_NUM=00
> E: ID_VENDOR=HGST_HDN
> E: ID_VENDOR_ENC=HGST\x20HDN
> E: ID_VENDOR_ID=152d
> E: MAJOR=8
> E: MD_DEVICE=md127
> E: MD_DEVNAME=helios4:Main
> E: MD_FOREIGN=yes
> E: MD_STARTED=no
> E: MINOR=32
> E: SUBSYSTEM=block
> E: TAGS=:systemd:
> E: USEC_INITIALIZED=553826637

winnuc:/var/log # udevadm info -n /dev/sdd

> P: /devices/pci0000:00/0000:00:14.0/usb2/2-1/2-1:1.0/host2/target2:0:0/2:0:0:2/block/sdd
> N: sdd
> S: disk/by-id/usb-HGST_HDN_726060ALE614_152D00539000-0:2
> S: disk/by-path/pci-0000:00:14.0-usb-0:1:1.0-scsi-0:0:0:2
> E: COMPAT_SYMLINK_GENERATION=2
> E: DEVLINKS=/dev/disk/by-id/usb-HGST_HDN_726060ALE614_152D00539000-0:2 /dev/disk/by-path/pci-0000:00:14.0-usb-0:1:1.0-scsi-0:0:0:2
> E: DEVNAME=/dev/sdd
> E: DEVPATH=/devices/pci0000:00/0000:00:14.0/usb2/2-1/2-1:1.0/host2/target2:0:0/2:0:0:2/block/sdd
> E: DEVTYPE=disk
> E: DONT_DEL_PART_NODES=1
> E: ID_BUS=usb
> E: ID_FS_LABEL=helios4:Main
> E: ID_FS_LABEL_ENC=helios4:Main
> E: ID_FS_TYPE=linux_raid_member
> E: ID_FS_USAGE=raid
> E: ID_FS_UUID=0c53ed8f-2ee6-c022-a0be-c9c7307a30bb
> E: ID_FS_UUID_ENC=0c53ed8f-2ee6-c022-a0be-c9c7307a30bb
> E: ID_FS_UUID_SUB=e2cfe83d-609e-b906-1075-f4288d0ddaec
> E: ID_FS_UUID_SUB_ENC=e2cfe83d-609e-b906-1075-f4288d0ddaec
> E: ID_FS_VERSION=1.2
> E: ID_INSTANCE=0:2
> E: ID_MODEL=726060ALE614
> E: ID_MODEL_ENC=726060ALE614\x20\x20\x20\x20
> E: ID_MODEL_ID=0567
> E: ID_PATH=pci-0000:00:14.0-usb-0:1:1.0-scsi-0:0:0:2
> E: ID_PATH_TAG=pci-0000_00_14_0-usb-0_1_1_0-scsi-0_0_0_2
> E: ID_REVISION=0125
> E: ID_SERIAL=HGST_HDN_726060ALE614_152D00539000-0:2
> E: ID_SERIAL_SHORT=152D00539000
> E: ID_TYPE=disk
> E: ID_USB_DRIVER=usb-storage
> E: ID_USB_INTERFACES=:080650:
> E: ID_USB_INTERFACE_NUM=00
> E: ID_VENDOR=HGST_HDN
> E: ID_VENDOR_ENC=HGST\x20HDN
> E: ID_VENDOR_ID=152d
> E: MAJOR=8
> E: MD_DEVICE=md127
> E: MD_DEVNAME=helios4:Main
> E: MD_FOREIGN=yes
> E: MD_STARTED=no
> E: MINOR=48
> E: SUBSYSTEM=block
> E: TAGS=:systemd:
> E: USEC_INITIALIZED=553856242

winnuc:/var/log # udevadm info -n /dev/sde

> P: /devices/pci0000:00/0000:00:14.0/usb2/2-1/2-1:1.0/host2/target2:0:0/2:0:0:3/block/sde
> N: sde
> S: disk/by-id/usb-HGST_HDN_726060ALE614_152D00539000-0:3
> S: disk/by-path/pci-0000:00:14.0-usb-0:1:1.0-scsi-0:0:0:3
> E: COMPAT_SYMLINK_GENERATION=2
> E: DEVLINKS=/dev/disk/by-path/pci-0000:00:14.0-usb-0:1:1.0-scsi-0:0:0:3 /dev/disk/by-id/usb-HGST_HDN_726060ALE614_152D00539000-0:3
> E: DEVNAME=/dev/sde
> E: DEVPATH=/devices/pci0000:00/0000:00:14.0/usb2/2-1/2-1:1.0/host2/target2:0:0/2:0:0:3/block/sde
> E: DEVTYPE=disk
> E: DONT_DEL_PART_NODES=1
> E: ID_BUS=usb
> E: ID_FS_LABEL=helios4:Main
> E: ID_FS_LABEL_ENC=helios4:Main
> E: ID_FS_TYPE=linux_raid_member
> E: ID_FS_USAGE=raid
> E: ID_FS_UUID=0c53ed8f-2ee6-c022-a0be-c9c7307a30bb
> E: ID_FS_UUID_ENC=0c53ed8f-2ee6-c022-a0be-c9c7307a30bb
> E: ID_FS_UUID_SUB=39dc936b-2cc0-0754-4a46-564f9fa3ca60
> E: ID_FS_UUID_SUB_ENC=39dc936b-2cc0-0754-4a46-564f9fa3ca60
> E: ID_FS_VERSION=1.2
> E: ID_INSTANCE=0:3
> E: ID_MODEL=726060ALE614
> E: ID_MODEL_ENC=726060ALE614\x20\x20\x20\x20
> E: ID_MODEL_ID=0567
> E: ID_PATH=pci-0000:00:14.0-usb-0:1:1.0-scsi-0:0:0:3
> E: ID_PATH_TAG=pci-0000_00_14_0-usb-0_1_1_0-scsi-0_0_0_3
> E: ID_REVISION=0125
> E: ID_SERIAL=HGST_HDN_726060ALE614_152D00539000-0:3
> E: ID_SERIAL_SHORT=152D00539000
> E: ID_TYPE=disk
> E: ID_USB_DRIVER=usb-storage
> E: ID_USB_INTERFACES=:080650:
> E: ID_USB_INTERFACE_NUM=00
> E: ID_VENDOR=HGST_HDN
> E: ID_VENDOR_ENC=HGST\x20HDN
> E: ID_VENDOR_ID=152d
> E: MAJOR=8
> E: MD_DEVICE=md127
> E: MD_DEVNAME=helios4:Main
> E: MD_FOREIGN=yes
> E: MD_STARTED=yes
> E: MINOR=64
> E: SUBSYSTEM=block
> E: TAGS=:systemd:
> E: USEC_INITIALIZED=554188919

winnuc:/var/log # ls -la /dev/disk/by-id/

> total 0
> drwxr-xr-x 2 root root 360 Sep 26 12:24 .
> drwxr-xr-x 8 root root 160 Sep 26 12:15 ..
> lrwxrwxrwx 1 root root   9 Sep 26 12:15 ata-ADATA_SU650_2L352911KN2W -> ../../sda
> lrwxrwxrwx 1 root root  10 Sep 26 12:15 ata-ADATA_SU650_2L352911KN2W-part1 -> ../../sda1
> lrwxrwxrwx 1 root root  10 Sep 26 12:15 ata-ADATA_SU650_2L352911KN2W-part2 -> ../../sda2
> lrwxrwxrwx 1 root root  10 Sep 26 12:15 ata-ADATA_SU650_2L352911KN2W-part3 -> ../../sda3
> lrwxrwxrwx 1 root root  10 Sep 26 12:15 ata-ADATA_SU650_2L352911KN2W-part4 -> ../../sda4
> lrwxrwxrwx 1 root root  11 Sep 26 12:24 md-name-helios4:Main -> ../../md127
> lrwxrwxrwx 1 root root  11 Sep 26 12:24 md-uuid-0c53ed8f:2ee6c022:a0bec9c7:307a30bb -> ../../md127
> lrwxrwxrwx 1 root root   9 Sep 26 12:24 usb-HGST_HDN_726060ALE614_152D00539000-0:0 -> ../../sdb
> lrwxrwxrwx 1 root root   9 Sep 26 12:24 usb-HGST_HDN_726060ALE614_152D00539000-0:1 -> ../../sdc
> lrwxrwxrwx 1 root root   9 Sep 26 12:24 usb-HGST_HDN_726060ALE614_152D00539000-0:2 -> ../../sdd
> lrwxrwxrwx 1 root root   9 Sep 26 12:24 usb-HGST_HDN_726060ALE614_152D00539000-0:3 -> ../../sde
> lrwxrwxrwx 1 root root   9 Sep 26 12:15 wwn-0x5000000000000000 -> ../../sda
> lrwxrwxrwx 1 root root  10 Sep 26 12:15 wwn-0x5000000000000000-part1 -> ../../sda1
> lrwxrwxrwx 1 root root  10 Sep 26 12:15 wwn-0x5000000000000000-part2 -> ../../sda2
> lrwxrwxrwx 1 root root  10 Sep 26 12:15 wwn-0x5000000000000000-part3 -> ../../sda3
> lrwxrwxrwx 1 root root  10 Sep 26 12:15 wwn-0x5000000000000000-part4 -> ../../sda4

**Comment by phillxnet** [comment]
@azilber Thanks for the outputs, more to follow later on this but first:

> The drives I pulled from my old Helios4 NAS

You have a Helios4, Nice. Shame the Marvell ARMADA® 388 (2 x A9 cores) is 32 bit ARM as we now support ARM64, but not 32bit ARM unfortunately, plus the 2 GB RAM is rather tight.

> which have an md RAID-5 with btrfs over it. I imported the pool into Rockstor.

## No mdraid support on data pools.

We only support mdraid on the system disk, and then only just. Data disks must be btrfs only, and ideally with no partitioning. No LVM or mdraid is 'understood' on the additional pools. The results of mdraid under the data pools is completely untested/undetermined; and primarily unsupported. But see below for the additional primary serial problem here.

## The udev serial question and the detailed enclosure.

Thanks for the enclosure info. From the udev outputs we can see what Rockstor has to work with. And as suspected it is identical to another we have seen, I'll link to the forum thread below. But in that thread I think I fathomed a way to support such devices but it will need a bit of back and forth if you are game? And support will be within certain limits: i.e. drive bay tracking, not drive tracking, which can work if the understanding is there on the user side. However in a linked form report below an identical controller was observed to not even have this degree of predictability/stability. And so this does not bode well for your enclosure, although it's firmware by differ in this respect.

---
```
DEVLINKS=...  /dev/disk/by-id/usb-HGST_HDN_726060ALE614_152D00539000-0:0
DEVNAME=/dev/sdb
ID_SERIAL=HGST_HDN_726060ALE614_152D00539000-0:0
ID_SERIAL_SHORT=152D00539000
```
---
```
DEVLINKS=/dev/disk/by-id/usb-HGST_HDN_726060ALE614_152D00539000-0:1
DEVNAME=/dev/sdc
ID_SERIAL=HGST_HDN_726060ALE614_152D00539000-0:1
ID_SERIAL_SHORT=152D00539000
```
---
```
DEVLINKS=/dev/disk/by-id/usb-HGST_HDN_726060ALE614_152D00539000-0:2
DEVNAME=/dev/sdd
ID_SERIAL=HGST_HDN_726060ALE614_152D00539000-0:2
ID_SERIAL_SHORT=152D00539000
```
---
```
DEVLINKS=... /dev/disk/by-id/usb-HGST_HDN_726060ALE614_152D00539000-0:3
DEVNAME=/dev/sde
ID_SERIAL=HGST_HDN_726060ALE614_152D00539000-0:3
ID_SERIAL_SHORT=152D00539000
```
---

And the abstracted by-id symlink names in /dev/disk/by-id

```
... usb-HGST_HDN_726060ALE614_152D00539000-0:0 -> ../../sdb
... usb-HGST_HDN_726060ALE614_152D00539000-0:1 -> ../../sdc
... usb-HGST_HDN_726060ALE614_152D00539000-0:2 -> ../../sdd
... usb-HGST_HDN_726060ALE614_152D00539000-0:3 -> ../../sde
```

So as can be seen, this enclosure has given all drives the ID_SERIAL_SHORT of 152D00539000. And from your excellent report we see this as the serial of the enclosure itself (abridged):

> Bus 002 Device 002: ID 152d:0567 JMicron Technology Corp. / JMicron USA Technology Corp. JMS567 SATA 6Gb/s bridge
> bcdUSB 3.00
> idVendor 0x152d JMicron Technology Corp. / JMicron USA Technology Corp.
> idProduct 0x0567 JMS567 SATA 6Gb/s bridge
> iManufacturer 10 JMicron
> iProduct 11 USB to ATA/ATAPI Bridge
> iSerial 5 152D00539000


This does not relate to any of the drives individual serial numbers and is the preferred serial for Rockstor (against the ID_SERIAL), bar some special arrangements made for the system disk only (in mdraid), and in the case of data disks LUKS abstractions. So we have 2 complications:

1. Mdraid under the data pool - The pool would need to be re-created using the bare disks and btrfs raid, no partitioning or any other layer 'in the way' between btrfs and the raw disk. This way we keep our moving parts to a bare minimum. Rockstor expect direct raw access to each disk and it's considered an ideal to minimise layers and to keep thinks as simple as possible.
2. This enclosure (it's controller) is obfuscating the drives.

## We have seen this enclosure behaviour before.
As suspected we have seen this type of enclosure obfuscation before. See the following forum thread for details.

https://forum.rockstor.com/t/usb-disks-with-no-serials/4365/6
and prior to that:

Models displaying this behaviour:

- 8 bay Orico unit (JMS567 controller) - forum member outicnz report
- 4 bay ORICO USB 3.0 - forum member beaglenz report
- 5 bay ORICO USB 3.0 - forum member Brett_Abela report

In that thread we have from forum member outicnz

"To answer your question on whether the “0:0” or “0:1” is assigned by the bay of the drive. Sadly not. The last digit sequentially increases for each drive detected regardless of the bay that it is located in."

If this is the case with your enclosure then it is not currently possible for Rockstor to track the disks as they will change name on each boot potentially.  And we expect by-id names to be stable as they were intended to be. These enclosure undermine that design. See from:

"Revise internal use and format of device names" #1320

> A World Wide Identifier (WWID) is a persistent and system independent identifier required of all SCSI standards compliant devices.
> Reference:-
> https://access.redhat.com/documentation/en-US/Red_Hat_Enterprise_Linux/7/html/Storage_Administration_Guide/persistent_naming.html

The very intention of by-id and wwid naming is to not change from boot to boot. These enclosures ensure that this is no longer the case.

Interesting that you found the smarmontools reports the serial in this case however. But we also depend on the by-id names being stable: hence very likely incompatible hardware.

Cautionary note on a potential work around. Udev assigns the affix of likely a usb subdevice or the like in these circumstances of "-0:0", "-0:1". and by prioritizing ID_SERIAL and not it's short variant in these 'special' cases it may be possible to throw an extreme amount of caution too-the-wind and have Rockstor be totally unaware / unable to know one drive from another between boots. This is highly un-advisable however. The caveat / exception required would be in the serial retrieval code. But the assumptions of Rockstor based on the persistence of these drive 'handles' in the above referenced recommendations from RedHad not actually being persistent is a major worry. In essence, every time the machine powers up, or as is the USB's bus often want, a reset is initiated. All the serials of all the drives will be re-arranged. This makes a mockery of any 'state' that Rockstor tries to track via these supposed 'persistent' names. That is the downside of obfuscating a member's otherwise hardwired serial number.

We now have a number of reports of these enclosures. I've not given up hope entirely but if, as in the other cases reported, the enclosure arbitrarily attributes, or causes to be arbitrarily attributed the "-0:0" affix to it's own serial to 'uniquely' identify those members then this is very much in the dodgy range of hardware. Shame as some of these are really quite nice.

Have a read of those forum threads and in the mean time I may end up adding this device's serial to our list of known 'dodgy' serials via the linked issue opened with a reference to your report. Also if you could identify for sure the stability or instability of those udev interpretations (the long serial) that may help to see if all of these controller implementations are unstable in that respect. I..e in your case does the smarmontools serial (real drive serial) stick to the same "-0:*" affix over reboots and power cycles and drive re-locations within the enclosure. Rockstor supports settings per drive, it identifies/tracks a drive via it's serial. Hence required stable serial. We assume udev does the same, consistently, as it, in turn, assumes serial stability. And there is the nub of the problem. Hardware obfuscation of this otherwise dependable reference.

**Comment by Hooverdan96** [comment]
@phillxnet do you want to close this, considering that there were at least 2 more PRs that have been merged some time ago (#2320 and #2345)?
